File: src/main/java/template_method/BasePostController.java
package template_method;

import template_method.exception.ApplicationException;
import template_method.exception.BadRequestException;
import template_method.exception.ErrorResponse;

public abstract class BasePostController<RQ extends RequestBody, RS extends ResponseBody, Dto> {

    public final Response<?> execute(RQ requestBody) {
        try {
            checkRequest(requestBody);
            return createResponse(doService(mapToDto(requestBody)));
        } catch (BadRequestException e) {
            return createErrorResponse(e.getMessage(), 400);
        } catch (ApplicationException e) {
            return createErrorResponse(e.getMessage(), e.getStatus());
        } catch (Exception e) {
            return createErrorResponse(e.getMessage(), 500);
        }
    }

    // Hook Method
    protected void checkRequest(RQ requestBody) {
    }

    protected abstract Dto mapToDto(RQ requestBody);

    protected abstract RS doService(Dto dto);

    protected abstract Response<RS> createResponse(RS responseBody);

    private Response<ErrorResponse> createErrorResponse(String message, int status) {
        return new Response<>(new ErrorResponse(message), status);
    }

}

File: src/main/kotlin/template_method/BasePostController.kt
package template_method

import template_method.exception.ApplicationException
import template_method.exception.BadRequestException
import template_method.exception.ErrorResponse

abstract class BasePostController<RQ : RequestBody, RS : ResponseBody, Dto> {
    fun execute(reqBody: RQ): Response<*> {
        return try {
            checkRequest(reqBody)
            createResponse(doService(mapToDto(reqBody)))
        } catch (e: BadRequestException) {
            createErrorResponse(e.message ?: "bad request", 400)
        } catch (e: ApplicationException) {
            createErrorResponse(e.message ?: "application error", e.status)
        } catch (e: Exception) {
            createErrorResponse(e.message ?: "unknown error", 500)
        }
    }

    // Hook Method
    open fun checkRequest(requestBody: RQ) {
    }

    protected abstract fun mapToDto(requestBody: RQ): Dto

    protected abstract fun doService(dto: Dto): RS

    protected abstract fun createResponse(responseBody: RS): Response<RS>

    private fun createErrorResponse(message: String, status: Int): Response<ErrorResponse> {
        return Response(ErrorResponse(message), status)
    }
}